### DISCUSSION TITLE
FLCAC Metadata Requirements

### DISCUSSION THREAD

**Discussion by dt-woods** [NETL-RIC/ElectricityLCI#239] [Q&A]
What are they?

**Comment by dt-woods** [comment]
@bl-young, there's been a discussion on our end whether we need to 'bulk up' the flow metadata in the baseline's JSON-LD. Currently, I'm using fedelemflowlist v.1.2.2 to map elementary flows to get their UUIDs; however, I'm seeing that the CAS number and the FEDEFL version numbers are not carried over. The Commons seems to have version 1.0 available. Any recommendations on paths forward to meet requirements and utilize exiting tools?

**Reply by bl-young** [reply]
The repository on the commons is out of date we are working to update that repo. Are these flows that you want to generate yourself within an eLCI json? Could you use the [write_jsonld()](https://github.com/USEPA/fedelemflowlist/blob/master/fedelemflowlist/__init__.py#L69) function to write all flows to zip based on the full list of elementary flows you need? This would basically create the flow objects in exactly the same way that we do for the full FEDEFL (and would also ensure you capture any new metadata we add in the future).

write_jsonld takes `flows` parameter which is a df in the flow list format, so you could take the flowlist using `fedelemflowlist.get_flows()` and query it for all of the flows in the eLCI output.

**Reply by dt-woods** [reply]
I was not aware of that! Thanks, Ben. I'll give it a try.
@m-jamieson, you recall any other metadata hangups with the electricity baseline for publishing on the Commons?

**Reply by m-jamieson** [reply]
Hangup might be a strong word.  I *think* that all the unit process metadata is there to satisfy FLCAC requirements...or at least they were satisfied at the time. 

Along the topic of flows though, we do create a number of product and waste flows (generated by RCRAInfo) that don't fit within the elementary flows list. There are no metadata associated with these, and unfortunately, I don't know what flow-level metadata requirements there are in the commons.

We should also probably move the waste flows from "Elementary flows/emission/waste" to "Technosphere Flows" or make a new "Waste Flows" category. Guess I'll call out @bl-young again here - has there been a standard set somewhere on waste flows as provided by StEWI?

**Reply by bl-young** [reply]
Yes RCRA... I believe we have these from USEEIO as well. When we turned that into JSON this is how it manifests:
https://www.lcacommons.gov/lca-collaboration/US_Environmental_Protection_Agency/USEEIO_v2/datasets/Flows/Waste%20Flows

Is that the best solution? I may need to take that back and think on it in terms of the specific categories/contexts for those flows. I'm open to adjusting the USEEIO repo for those flows if that makes sense and you all have a better solution.

I do recommend using the UUID function in esupy ([here](https://github.com/USEPA/esupy/blob/main/esupy/util.py#L70)] for consistent UUID generation. It would be good if the same RCRA flows always had the same UUID and contexts across repos.

I also seem to recall the existing baseline having some additional elementary flows that arent in the FEDEFL, maybe they were some water flows? Happy to set something up to talk through this in more detail and/or review what you have so far if that's useful.

I'll also add I have been working on creating a schema for Sources through esupy (https://github.com/USEPA/esupy/pull/51) if thats valuable for you.

**Reply by bl-young** [reply]
@dt-woods in https://github.com/USEPA/LCIAformatter/issues/103 I've been wanting to use the existing function in fedelemflowlist to write flows in lciafmt to make them more consistent. I started that with https://github.com/USEPA/LCIAformatter/commit/36a37c417e65470e96740dde889e3735554a51b7 and https://github.com/USEPA/fedelemflowlist/commit/7ed509adc0153ad5c6f5f337af54483ea769ca79. I realized that the way things are currently set up, it works best to pass the `zipWriter` object through to fedelemflowlist. Anyhow, I don't recall exactly how its set up in eLCI, but perhaps somewhere in [_save_to_json](https://github.com/KeyLogicLCA/ElectricityLCI/blob/development/electricitylci/olca_jsonld_writer.py#L1712) you could leverage `fedelemflowlist.write_jsonld()` for the elementary flows. Let me know if you give it a shot. I'm happy to work through additional changes within fedelemflowlist

**Reply by dt-woods** [reply]
I'm pretty sure our UUID generation is the same as esupy's, but good to know that we would could (maybe even should) replace it with an import to that method. It assumes that the args we send the method are consistent.
https://github.com/KeyLogicLCA/ElectricityLCI/blob/8c5c5bb1c783b27ecef79cd4fa5e86326ac39021/electricitylci/olca_jsonld_writer.py#L1858

**Comment by dt-woods** [comment]
Here's my proposed solution, which happens within olca_jsonld_writer.py:

```py
def _flow(dict_d, flowprop, dict_s):
    """Generate a reference to a flow object.

    Called by :func:`_exchange`.

    Parameters
    ----------
    dict_d : dict
        Flow data dictionary.
    flowprop : olca_schema.FlowProperty or olca_schema.Ref
        FlowProperty or Ref to a FlowProperty object.
    dict_s : dict
        Dictionary with olca_schema root entities.

    Returns
    -------
    tuple
        olca_schema.Ref : Reference to a flow object.
        dict : Updated dictionary with olca_schema root entities.
    """
    if not isinstance(dict_d, dict):
        logging.warning("No flow data received!")
        return (None, dict_s)

    uid = _val(dict_d, 'id', '@id')
    name = _val(dict_d, 'name')
    category_path = _val(dict_d, 'category', default='')
    is_waste = "waste" in category_path.lower()

    # HOTFIX: remove technosphere/3rd party flow check;
    # it duplicates every waste flow in the JSON-LD [2023-12-05; TWD]

    # Check for flow existence
    if uid in dict_s['Flow']['ids']:
        idx = dict_s['Flow']['ids'].index(uid)
        flow = dict_s['Flow']['objs'][idx]
        logging.debug("Found previous flow, '%s'" % flow.name)
    else:
        logging.debug("Creating new flow for, '%s' (%s)" % (name, uid))
        if _uid_is_valid(uid, 3) or _uid_is_valid(uid, 4):
            # Keep the good UUID
            pass
        else:
            # Generate new v3 ID based on standard format
            logging.debug("Generating new UUID for flow, '%s'" % name)
            uid = _uid(o.ModelType.FLOW, category_path, name)

        # Correct the default flow type for waste flows.
        def_type = "ELEMENTARY_FLOW"
        if is_waste:
            dict_d['flowType'] = "WASTE_FLOW"
            def_type = "WASTE_FLOW"

        f_type = _flow_type(_val(dict_d, 'flowType', default=def_type))
        flow = o.new_flow(name, f_type, flowprop)

        # Get FEDEFL version number and CAS number
        fed_version = _get_fedefl_version()
        fed_cas, fed_form, fed_syn = _get_fedefl_meta(uid)

        # Hotfix: Add missing flow metadata.
        # https://github.com/USEPA/ElectricityLCI/discussions/239
        flow.id = uid
        flow.category = category_path
        flow.description = "Based on FEDELEMFLOW version %s" % fed_version
        flow.cas = fed_cas
        flow.formula = fed_form
        flow.synonyms = fed_syn

        # Update master list
        dict_s['Flow']['ids'].append(uid)
        dict_s['Flow']['objs'].append(flow)
    return (flow.to_ref(), dict_s)


def _get_fedefl_meta(flow_id):
    """Query the FEDEFL for flow metadata (i.e., CAS No, formula, and synonyms)

    Parameters
    ----------
    flow_id : str
        Flow universally unique identifier (mapped to FEDEFL).

    Returns
    -------
    tuple
        A tuple of length three:

        - str: CAS No
        - str: Formula
        - str: Synonyms
    """
    # Initialize return values
    f_cas = ""
    f_form = ""
    f_syn = ""

    # Read the FEDEFL into memory
    flowlist = fedelemflowlist.get_flows(False)

    # Query the flow ID and count rows that match
    tmp_df = flowlist.query("`Flow UUID` == '%s'" % flow_id)
    num_matches = len(tmp_df)

    # Query results for metadata.
    if num_matches == 0:
        logging.info("Failed to find FEDEFL flow (%s)!" % flow_id)
    elif num_matches == 1:
        # Convert data frame to series
        tmp_df = tmp_df.iloc[0]
        f_cas = tmp_df['CAS No']
        f_form = tmp_df['Formula']
        f_syn = tmp_df['Synonyms']
    else:
        logging.warning(
            "Found %d matches for FEDEFL flow (%s); using first instance." % (
                num_matches, flow_id)
        )
        tmp_df = tmp_df.iloc[0]
        f_cas = tmp_df['CAS No']
        f_form = tmp_df['Formula']
        f_syn = tmp_df['Synonyms']

    return (f_cas, f_form, f_syn)


def _get_fedefl_version():
    """Return the version string of Federal Elementary Flow List

    Currently hard-coded to return version 1.0.0 as defined in USEPA's
    LCIAfmt. See https://github.com/USEPA/fedelemflowlist/issues/199.

    Returns
    -------
    str
        Version string (e.g., '1.0.0' or 'n/a' if not available)
    """
    try:
        # Wishy thinking
        return fedelemflowlist.__version__
    except AttributeError:
        return "1.0.0"
```

**Reply by bl-young** [reply]
for `_get_fedefl_version()`, (pending https://github.com/USEPA/fedelemflowlist/issues/199) could you instead use:
```
fedelemflowlist.globals.flow_list_specs['list_version']
```

Though really, I'll caution that there is a distinction between the version of the package and the version of the flowlist (you could have the latest version of fedelemflowlist but you might have an older version of the FlowList data file (parquet) stored locally which could get used).

**Reply by dt-woods** [reply]
Good to know! Yeah, I want the version of the flow list, not the Python package. I'll incorporate the code you sent. Thanks, Ben!

**Reply by dt-woods** [reply]
What I'm seeing as a means to getting the file version is the following:

```py
from esupy.processed_data_mgmt import find_file
import fedelemflowlist

def _get_fedefl_version():
    """Return the version string of local Federal Elementary Flow List

    Returns
    -------
    str
        Version string. Defaults to '1.0.0' if methods fail.
    """
    version = None
    try:
        # Adapted from:
        # https://github.com/USEPA/ElectricityLCI/discussions/239
        meta = fedelemflowlist.globals.set_metadata()
        fpath = fedelemflowlist.globals.fedefl_path

        # From esupy's documentation: returns the most recent version of the
        # file requested---so does not have to match ``meta.tool_version``.
        f = find_file(meta, fpath)
        if f:
            # The parquet file name is based on ``set_metadata`` function 
            # in fedelemflowlist's globals.py.
            r = re.compile(".*_v(\d+\.\d+\.\d+)_.*")
            m = r.match(f.name)
            if m:
                version = m.group(1)
        if version is None:
            version = fedelemflowlist.globals.flow_list_specs['list_version']
    except AttributeError:
        version = "1.0.0"
    
    return version
```

**Reply by bl-young** [reply]
Moving forward we will be distinguishing the package version from the list version (see https://github.com/USEPA/fedelemflowlist/discussions/205). Did you end up implementing this function above as I think it could be useful to move into fedelemflowlist itself.

**Comment by dt-woods** [comment]
Running `flowlist = fedelemflowlist.get_flows(False)` for each flow is laughably slow and litters the info log with thousands of `processed_data_mgmt:load_preprocessed_output:Returning ~/fedelemflowlist/FedElemFlowListMaster_v1.2.0_e57a542.parquet`

**Reply by bl-young** [reply]
see my proposed solution in https://github.com/KeyLogicLCA/ElectricityLCI/pull/12

**Comment by dt-woods** [comment]
Despite the slowness, it gets the job done.
<img width="977" alt="Screenshot 2024-08-29 at 15 29 54" src="https://github.com/user-attachments/assets/b7f6f12f-86da-4884-ace9-c9415752631e">

**Comment by dt-woods** [comment]
I think it would be a benefit to move this step to the post processing phase.

**Comment by dt-woods** [comment]
How do I know what version of the parquet file I have?

**Reply by bl-young** [reply]
there should be a metadata file (json) that is downloaded as well that indicates that. Also the name of the file should include the version number.

**Comment by dt-woods** [comment]
# Process Metadata YAML
Here's a little work on figuring out what is assigned where.

The metadata keys:

- **'biomass'**
    - Should be assigned to Electricity - BIOMASS processes, but is the default: replace_egrid data treatment and data selection. The biomass process description is not set (uses the default defined in olcaschema_genprocess in generation.py).
-  'coal_transport_upstream'
    - Correctly assigned to 'coal transport' processes under 2121: Coal Mining
-  'coal_upstream'
    - Correctly assigned to 'coal extraction and processing' processes under 2121: Coal Mining
-  'construction_upstream'
    - Correctly assigned to 'power plant construction - coal_const' and 'power plant construction - ngcc_const' under 2371: Utility System Construction
-  **'consumption_mix'**
    - Should be assigned to 'at grid; consumption mix'
-  'default'
    - The default process description, sampling procedure, intended application, dataset owner/generator/documentor, access restrictions, project description, LCI method, completeness, data treatment, data selection, data collection period, reviewer, and sources
    - **The project description may be outdated for version 2.0**
-  'default_boundary_statement'
    - Used to define sampling procedure in 'default'
-  'default_datacollection_statement'
    - Used to define sampling procedure in 'default'
-  'default_uncertainty_statement'
    - Used to define sampling procedure in 'default'
-  'gas_upstream'
    - Correctly assigned to 'natural gas extraction and processing' processes under 2212: natural gas distribution
-  **'generation_mix'**
    - The process description is not used, but the data treatment and data collection period appear to match. 
-  'geothermal'
    - The process description is not used, but the data treatment and data selection appear correct.
    - The sampling procedure is the default SamplingProcedure
    - Data collection period is correct
    - Sources are correct
-  **'hydro'**
   - All appears to be default; not used
-  'no_uncertainty_statement'
    - Used to define default sampling procedure 
-  'nuclear_upstream'
    - Correctly assigned to 'nuclear fuel extraction, processing, and transport' under 3251: Basic Chemical Manufacturing 
    - The sampling procedure is set to default
-  'oil_upstream'
    - Correctly assigned to 'petroleum extraction and processing' processes under 3241: Petroleum and Coal Products Manufacturing
-  **'solar'**
    - Correctly assigns data treatment, data selection, data collection period, and sources.
    - Default samling procedure and process description taken from olcaschema_genprocess
-  'solarpv_construction_upstream'
    - Correctly assigns the process description, dataset owner, data treatment, data selection, data collection period
    - The sampling procedure is from default. 
-  'solartherm_construction_upstream'
    - Correctly assigns process description, data treatment, data selection, data collection period, dataset owner/generator/documentor
    - The sampling procedure is from default.
-  'solarthermal'
    - Everything except process description and sampling procedure appears correct.
-  'source_ingwersen_2024'
    - Used to reference Ingwersen et al. (in prep) manuscript.
-  'source_netl_coal_baseline_2024'
    - Used to reference NETL's coal baseline report
-  'wind'
    - Everything except process description and sampling procedure.
-  'wind_construction_upstream'
    -  Everything except sampling procedure.

**Reply by dt-woods** [reply]
Several of these issues are solved thanks to correcting the `sampling_procedures` to `SamplingProcedures` in process_metadata.yml.

**Reply by dt-woods** [reply]
Notably, the YAML documentation is parsed using `process_doc_creation` and `process_description_creation` found in process_dictionary_writer.py.

In upstream_dict.py, the fuel categories are used to find the appropriate entry in process_metadata.yml (by appending "_upstream" to the fuel category name).

**Reply by dt-woods** [reply]
Found the potential problem with missing fuel categories in VALID_FUEL_CATS here (no biomass and no hydro):
https://github.com/USEPA/ElectricityLCI/blob/d4cfd8951135b52d519c26ecf6dc0e0f3d8f00db/electricitylci/process_dictionary_writer.py#L165
